adds onQuery option which is called everytime before a fetch happens

Jeffrey-Zutt opened the pull request:
### Description

This PR adds an `onQuery` option to `useQuery()`. 
Devs can opt to add a callback to handle side-effects related to the query in a similar fashion as `useMutation()` `onMutate`.

Example:
```tsx
const result = useQuery(
   queryKey,
  () => 'data',
  { onQuery: () => console.log('onQuery') }
)
```

Also:
Great job on react-query! Love using it :-)

daraselia commented on docs/src/pages/reference/useQuery.md:
```diff
@@ -127,6 +128,9 @@ const result = useQuery({
   - If set to `['isStale']` for example, the component will not re-render when the `isStale` property changes.
+- `onQuery: () => void`
+  - Optional
+  - This function fire will any time right before the query tries to fetch new data.
```
`...will fire...`

Jeffrey-Zutt replied:
Thanks :-)

daraselia commented on src/core/retryer.ts:
```diff
@@ -6,6 +6,7 @@ import { sleep } from './utils'
 
 interface RetryerConfig<TData = unknown, TError = unknown> {
   fn: () => TData | Promise<TData>
+  onQuery?: () => void
```
Would it be useful to pass query time variables as an argument to callback function?

TkDodo commented:
How will this work for infinite queries please? When a refetch happens, will it be called once, or once for each page that is refetched?

Jeffrey-Zutt commented:
Hi TkoDodo, it will be called for every page that is fetched

TkDodo commented:
> Hi TkoDodo, it will be called for every page that is fetched

cool, can you add a test for this maybe? Also, should we get access to the `pageParams` or the `index` or so ? it also needs documentation please :)

Jeffrey-Zutt commented:
Yes, of course :-).

- Added a test for `useInfiniteQuery`
- onQuery is now called with a `QueryFunctionContext`.
    - updated tests accordingly  
- Updated documentation

Please let me know if you want anything changed!

TkDodo commented on src/core/types.ts:
```diff
@@ -62,6 +62,11 @@ export interface QueryOptions<
+   * This callback will fire right before the query will start fetching new data.
+   * Useful for side-effects that are specific to this query.
+   */
+  onQuery?: (context:QueryFunctionContext) => void
```
should this not be defined on the `QueryObserverOptions`, like the other callbacks (`onSuccess`, `onError`, `onSettled`) ? We want to be able to have different callbacks per observer

Jeffrey-Zutt replied:
Thank you! These are valid points.

- I moved `onQuery` to `QueryObserverOptions`.
- When `onQuery` exists,  we're calling it from `queryFn()` now, instead of the retryer.
    - It makes the code a lot simpler: no need to modify infiniteQueryBehavior anymore.

TkDodo commented on src/core/infiniteQueryBehavior.ts:
```diff
@@ -8,7 +8,43 @@ export function infiniteQueryBehavior<
+
     onFetch: context => {
+      context.onQueryFn = () => {
+        const fetchMore = context.fetchOptions?.meta?.fetchMore
+        const pageParam = fetchMore?.pageParam
+        const oldPages = context.state.data?.pages || []
+        const isFetchingNextPage = fetchMore?.direction === 'forward'
+        const isFetchingPreviousPage = fetchMore?.direction === 'backward'
+
+        let param: unknown|undefined;
+
+        // first page
+        if (!oldPages.length) {
+          param = undefined
+        }
+        // Fetch next page?
+        else if (isFetchingNextPage) {
+          const manual = typeof pageParam !== 'undefined'
+          param = manual
+            ? pageParam
+            : getNextPageParam(context.options, oldPages)
+        }
+        // Fetch previous page?
+        else if (isFetchingPreviousPage) {
+          const manual = typeof pageParam !== 'undefined'
+          param = manual
+            ? pageParam
+            : getPreviousPageParam(context.options, oldPages)
+        }
+
+        const queryFnContext: QueryFunctionContext = {
+          queryKey: context.queryKey,
+          pageParam: param
+        }
+
+        context.options.onQuery?.(queryFnContext)
```
- none of this needs to be done if `context.options.onQuery` is `undefined`
- i see a bit of an overlap with what we are doing in `context.fetchFn`. not sure if it's good to call `getNextPageParam` / `getPreviousPageParam` twice - once for `onQuery` and once for the actual fetch. I'm not sure here so I'd like to wait on more feedback from either @tannerlinsley or @boschni

boschni replied:
Hi Jeffrey! Think this can also be implemented in the `QueryObserver.onQueryUpdate` method. There you should be able to listen for the fetch action and trigger the callback. Then it is consistent with the other callbacks and it also works outside of the hooks. Regarding the naming, maybe `onFetch` would be a bit more descriptive? Then it also corresponds more with the `fetchFn` option.

Jeffrey-Zutt replied:
Please see previous comment, this whole code-block is unnecessary because `onQuery` is now called from `queryFn` .

boschni replied:
My comment was not really related to this piece of code, just placed it here because I got tagged ;)

TkDodo replied:
> maybe onFetch would be a bit more descriptive?

I like both `onQuery` as well as `onFetch` 🤔 . `onQuery` for `useQuery` corresponds well with `onMutate` for `useMutation`.`onFetch` would also make sense because we have the `isFetching` flag 🤷

boschni replied:
Yeah it also relates to the `refetch()` method in `useQuery`. When doing mutations you call `mutate()` so it makes sense to use `onMutate` there.

TkDodo replied:
if you prefer `onFetch`, then let's rename 👍

Jeffrey-Zutt replied:
Thanks guys! I'll rename to `onFetch` and move the call to `QueryObserver.onQueryUpdate`.

Jeffrey-Zutt replied:
@TkDodo @boschni 
Update:

- I renamed the function `onFetch`.
- The function is called from the `queryObserver`

Let me know what you think :-).

Jeffrey-Zutt replied:
@TkDodo @boschni 
- I renamed the function `onFetch`
- I moved the call to `QueryObserver.onQueryUpdate`

Let me know what you think :-)

TkDodo reviewed (changes requested):
I think it's good 👍 , but need some more feedback. I would also appreciate another test that shows that `onQuery` is called even when a `retry` happens, maybe we should even mention this in the docs? Because if an error happens, with default of `retry: 3`, the function would be called 4 times before we get to an error state (unless I misinterpreted the code 😅 )

Jeffrey-Zutt commented:
> I would also appreciate another test that shows that `onQuery` is called even when a `retry` happens, maybe we should even mention this in the docs? Because if an error happens, with default of `retry: 3`, the function would be called 4 times before we get to an error state (unless I misinterpreted the code 😅 )

I added the test and updated the docs.
Please let me know what you think :-)

boschni commented:
Nice! Two question that come to mind:

1. Should the `onFetch` function be called once, or also on each retry? What is its purpose?

2. The functionality can also be achieved with `useEffect`, do we need a separate callback?

```js
const { isFetching } = useQuery()

useEffect(() => {
  if (isFetching) {
    // Run code...
  }
}, [isFetching])
```

TkDodo commented:
> The functionality can also be achieved with useEffect, do we need a separate callback?

can't all of the callbacks theoretically be achieved with useEffect?

if we include retries and call it once per retry, I don't think that we can actually do it with useEffect :)

Jeffrey-Zutt commented:
@boschni 

1. We'd like onFetch to be called on every retry as well. For instance, we'd like to able to log query-duration for client-side analytics. As @TkDodo says, that's currently not possible using useEffect with isFetching.

2. For consistency/simplicity I think it's nice to be able to configure these callbacks in a similar way as `useMutation`.
I think it's easier to configure them 'from top', instead of relying on its result to hook into it.

E.g.

```tsx
const useMyQuery = (enableLogging:boolean) => {
   return useQuery(
      key, 
      () => { /* fetch */ }, 
      enableLogging 
            ? configForLogging
            : undefined
)}
```

Instead of:

```tsx
const useMyQuery = (enableLogging:boolean) => {
   const { isFetching, ...rest } = useQuery(
      key, 
      () => { /* fetch */ }
   )
   
   // I don't like this as tightly coupled here.
   React.useEffect(() => {
      if (enableLogging) {
         if (isFetching) {
            startClock()
         } else {
            stopClock()
         }
      }
      
   }, [ isFetching, enableLogging ])
		
   return {
      isFetching,
      ...rest
   }
)}
```

I think you're right though. We could probably solve it using some well chosen `useEffect()`, or some other way.
I also think the onFetch is a nice addition for consistency :-).

boschni commented:
I'm not sure if this would be the best approach for request logging because if you have multiple components using the same query then it would be registered multiple times. Usually I place functionality like this (authorization/logging) in an API client:

```js
class SomeApi {
  fetchSomething() {
     this.request("/something")
  }
  request(url) {
    log(url)
  }
}

const api = new SomeApi()

function Component() {
  const result = useQuery(() => api.fetchSomething())
}
```

tannerlinsley reviewed (commented):
I personally don’t find this addition helpful considering this functionality is easily achieved in user land by either augmenting or proxying the query function with the desired additional fetch logic. It doesn’t add any control or configuration to the internal lifecycle that is unique beyond being a “hook” into the fetch function, which already exist in user land anyway.

tannerlinsley commented:
Just to make sure none of this was in vain, I would gladly accept a PR to improve the documentation on this point. It's much easier to educate people that they can write their own "onFetch" functionality by simply using their own function that's passed to useQuery. This pattern offers more control, is an API decision that inverts control and power back to the user instead of drawing it away into abstraction, and will arguably help any dev that implements the pattern become more skilled in utilizing the general pattern of inversion of control to achieve domain-specific results.
